src/Autofac/Util/TypeExtensions.cs

```diff
@@ -182,6 +182,45 @@ public static bool IsGenericTypeDefinedBy(this Type @this, Type openGeneric)
                     && key.Item1.GetGenericTypeDefinition() == key.Item2);
         }
 
+        /// <summary>
+        /// Checks whether this type is an open generic type of a given type.
+        /// </summary>
+        /// <param name="this">The type we are checking.</param>
+        /// <param name="type">The open generic type to validate against.</param>
+        /// <returns>True if <paramref name="this"/> is a closed type of <paramref name="type"/>. False otherwise.</returns>
+        public static bool IsOpenGenericTypeOf(this Type @this, Type type)
```

Comment:
I think I got confused by this method because `this` is a closed generic and `type` is the open generic, so the usage would be like `typeof(IGeneric<string>).IsOpenGenericTypeOf(typeof(IGeneric<>)`... but that's backwards for the name. It should be `IsClosedGenericTypeOf` if it's that direction, right? `X.IsClosedGenericTypeOf(openGeneric)`

It would be really good to get some unit tests on this. I don't know if we have too many tests for these sorts of utilities, but for stuff like `IsAssignableFrom` vs `IsAssignableTo` it's always hard to remember the direction it operates, so seeing unit tests helps explain the intent.

It might also help in refactoring. For example, I see on line 198 that we're already checking if `@this == type` but then on line 203 there's the same thing. It's nice to be able to safely refactor and trust the tests will catch issues.

Reply:
Thanks @tillig  I've just refactored and added tests for this method, could you check it ?

Reply:
@RaymondHuy; just passing through, but I wonder if there should be some negative tests for `IsOpenGenericTypeOf` as well as positive ones?

Reply:
Thanks @alistairjevans  I've just added negative tests for it but not sure is it enough to cover ? 😕

Reply:
Looks like enough, thanks!